src/Support/SelectFields.php

```diff
@@ -43,14 +52,65 @@ public function __construct(ResolveInfo $info, GraphqlType $parentType, array $a
             $parentType = $parentType->getWrappedType(true);
         }
 
-        if (! is_null($info->fieldNodes[0]->selectionSet)) {
-            self::$args = $args;
+        $this->info = $info;
+        self::$args = $args;
+        $fields = self::getSelectableFieldsAndRelations($this->getFieldSelection(5), $parentType);
+        $this->select = $fields[0];
+        $this->relations = $fields[1];
+    }
+
+    public function getFieldSelection($depth = 0)
+    {
+        $data = [];
+
+        if (! is_null($this->info->fieldNodes[0]->selectionSet)) {
+            /** @var FieldNode $fieldNode */
+            foreach ($this->info->fieldNodes as $fieldNode) {
+                if (! is_null($fieldNode->selectionSet)) {
+                    $data = array_merge_recursive($data, $this->foldSelectionSet($fieldNode->selectionSet, $depth));
+                }
+            }
+        }
 
-            $fields = self::getSelectableFieldsAndRelations($info->getFieldSelection(5), $parentType);
+        $fields['fields'] = $data;
+        $fields['args'] = self::$args;
 
-            $this->select = $fields[0];
-            $this->relations = $fields[1];
+        return $fields;
+    }
+
+    private function foldSelectionSet(SelectionSetNode $selectionSet, $descend)
```

Comment:
This whole logic shouldn't be intermixed within this class.

Extracting the field/argument datastructure is a separate concern.